gnd/src/main/java/com/google/android/gnd/persistence/mbtiles/MbtilesFootprintParser.java

```diff
@@ -110,20 +68,35 @@
     return result;
   }
 
-  private static List<JSONObject> getFeaturesArray(String jsonString) {
+  private Single<ImmutableList<TileSetJson>> getJsonTileSets(File jsonSource) {
     try {
-      // TODO: Separate parsing and intersection checks, make asyc (single, completable).
-      JSONObject geoJson = new JSONObject(jsonString);
-      // TODO: Make features constant.
-      return toArrayList(geoJson.getJSONArray(FEATURES_KEY));
-    } catch (JSONException e) {
-      Timber.e(e, "Unable to parse JSON");
+      String fileContents =
+          FileUtils.readFileToString(jsonSource, Charset.forName(JSON_SOURCE_CHARSET));
+      JSONObject geoJson = new JSONObject(fileContents);
+      JSONArray features = geoJson.getJSONArray(FEATURES_KEY);
+
+      ImmutableList<TileSetJson> tilesets =
+          stream(toArrayList(features)).map(TileSetJson::new).collect(toImmutableList());
+
+      return Single.just(tilesets);
+    } catch (Exception e) {
+      return Single.error(e);
     }
-    return ImmutableList.of();
   }
 
-  public ImmutableList<TileSetJson> getGeoJsonTiles(String jsonString) {
-    return stream(getFeaturesArray(jsonString)).map(TileSetJson::new).collect(toImmutableList());
+  public Single<ImmutableList<TileSet>> allTiles(File file) {
+    return getJsonTileSets(file).map(mapAndRecollect(this::jsonToTileSet)).doOnError(Timber::e);
+  }
+
+  /**
+   * Returns the immutable list of tiles specified in {@param geojson} that intersect {@param
+   * bounds}.
+   */
+  public Single<ImmutableList<TileSet>> intersectingTiles(LatLngBounds bounds, File file) {
+    return getJsonTileSets(file)
+        .map(filterAndRecollect(tile -> tile.boundsIntersect(bounds)))
```

Comment:
Perhaps `compose()`  would be better suited for such higher-order functions?

Reply:
@gino-m can you clarify a bit here? Do you mean `rx.compose` or `Function.compose`?

RX compose would be nice....if it were generic on on observables, but it's not afaict. One needs to pass an explicit `XTransformer` type to `compose` in which case the signature for these helpers would need to change to `SingleTransformer....` which would reduce the utility of these helpers a lot imo. I may be missing some way to make it generic using java's generics system though.